### ISSUE TITLE
Make code output stable across multiple runs

### ISSUE THREAD

**Issue by dktapps** [wol-soft/php-json-schema-model-generator#41]
**Is your feature request related to a problem? Please describe.**
Currently, this library relies on uniqid() quite a lot to ensure uniqueness in generated code. This is quite annoying when the generated code is in a git repository, because schema code changes every time my build script runs even if the schema itself was untouched.

**Describe the solution you'd like**
For class names, I've had some success with a custom class name generator that suffixes with `md5(json_encode(propertySchema))` instead of `uniqid()`. However, there's some additional places that I can't touch without modifying the library itself:

- [ArrayItemValidator](https://github.com/wol-soft/php-json-schema-model-generator/blob/master/src/Model/Validator/ArrayItemValidator.php#L47) generates variable names using uniqid(). I'm not actually very clear why this is needed at all, since validators are anyway split up into separate methods per property.
- [AbstractPropertyProcessor](https://github.com/wol-soft/php-json-schema-model-generator/blob/master/src/PropertyProcessor/Property/AbstractPropertyProcessor.php#L158) uses uniqid() to generate class names for rendering dependencies. Again, I'm not very clear why this is needed - isn't `ClassNameGenerator` enough to take care of this?

**Additional context**
![image](https://user-images.githubusercontent.com/14214667/136208232-6130476b-aa83-4288-a3fe-63c8b392b113.png)

**Comment by dktapps** [comment]
For the `ArrayItemValidator` case it seems like `md5(propertyName)` is sufficient. I checked nested arrays and their property names are `item of {actualPropertyName}`, so this seems fine.

**Comment by wol-soft** [comment]
I've had a quick view: for the `AbstractPropertyProvider` I think the uniqid is not necessary, as you already stated in the issue description.

The `ArrayItemValidator` on the other hand currently needs the suffix. You're right, each property is evaluated in a single function so conflicts won't occur. But if a property contains a nested array the validation for the property will contain nested array validators in a single function which will collide and overwrite each others invalid items collection (if the variable names are identical, eg. both use `$invalidItems`).

Consequently some kind of 'nested level counter' or something similar would be required to generate a reproducable suffix for the variable names. I'll have a look into it the next days.

For all my usages of the library, I don't commit the generated code (as it's just a duplication in my opinion. All information is already committed in the JSON-files used to generate the classes). I run my generate-script as a build step of the applications (like `composer install` to not commit the vendor directory). Is there a specific reason why you commit the generated code?

**Comment by wol-soft** [comment]
> I did some experimentation and found that a hash of this seems to be sufficient for nested arrays.

Yes, you are absolutely right. Clean and simple solution. I've added it in https://github.com/wol-soft/php-json-schema-model-generator/commit/1b7a21acb5e6c6e10dbd6d7c2e7143becce939f1 to the master as well as a patch for the `AbstractPropertyProvider`. You may test it with the latest master version.

Thanks for the hint 😃

**Comment by dktapps** [comment]
Thanks. Only thing that remains (AFAIK) is the ClassNameGenerator, but I solved that with my code using a custom impl that just md5's the contents of the json schema. There's probably a better way to do it though (is there a way to get the full property path in there?).

**Comment by wol-soft** [comment]
Yes, using this library in a library context currently has no proper solution. Maybe it's possible to wrap the generate script into a [Composer plugin](https://getcomposer.org/doc/articles/plugins.md#plugin-package). I've not yet tested this idea.
